plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+            if(red == 0)
+               continue;
+            co_await stream.async_write_some(false, boost::asio::buffer(buff, red));
+         }
       }
-
-      assert(current_request);
-      assert(std::holds_alternative<state_history::get_blocks_request_v0>(*current_request) ||
-             std::holds_alternative<state_history::get_blocks_request_v1>(*current_request));
-
-      std::visit(eosio::chain::overloaded{
-                    [&](eosio::state_history::get_blocks_request_v0& request) {
-                       state_history::get_blocks_result_v0 result;
-                       result.head = head;
-                       send_update(request, false, std::move(result), block, id); },
-                    [&](eosio::state_history::get_blocks_request_v1& request) {
-                       state_history::get_blocks_result_v1 result;
-                       result.head = head;
-                       send_update(request, request.fetch_finality_data, std::move(result), block, id); } },
-                 *current_request);
-   }
-
-   void send_update(const chain::signed_block_ptr& block, const chain::block_id_type& id) override {
-      if (no_request_or_not_max_messages_in_flight()) {
-         session_mgr.pop_entry(false);
-         return;
+      else {
+         co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(false)));
       }
-
-      auto block_num = block->block_num();
-      to_send_block_num = std::min(block_num, to_send_block_num);
-      send_update(state_history::block_position{block_num, id}, block, id);
    }
 
-   void send_update(bool changed) override {
-      if (changed || need_to_send_update) {
-         send_update(plugin.get_block_head(), nullptr, chain::block_id_type{});
-      } else {
-         session_mgr.pop_entry(false);
-      }
-   }
+   boost::asio::awaitable<void> write_loop() {
+      co_await readwrite_coro_exception_wrapper([this]() -> boost::asio::awaitable<void> {
+         get_status_result_v1 current_status_result;
+         struct block_package {
+            get_blocks_result_base blocks_result_base;
+            bool is_v1_request = false;
+            chain::block_num_type this_block_num = 0; //this shouldn't be needed post log de-mutexing
+            std::optional<locked_decompress_stream> trace_stream;
+            std::optional<locked_decompress_stream> state_stream;
+            std::optional<locked_decompress_stream> finality_stream;
+         };
+
+         while(true) {
+            if(!stream.is_open())
+               break;
+
+            std::deque<bool>             status_requests;
+            std::optional<block_package> block_to_send;
+
+            ///TODO: How to set main thread priority?
+            auto& self = *this; //gcc10 ICE workaround wrt capturing 'this' in a coro
+            co_await boost::asio::co_spawn(app().get_io_service(), [&]() -> boost::asio::awaitable<void> {
```

Comment:
Equivalent commentary here as with the read loop.  Naming the lambda or adding a comment block would make it crystal clear that this construct is for moving data from the main thread to this thread.